functional/lenses.md

```diff
@@ -0,0 +1,481 @@
+{
+    iterator.map(|r| r.as_ref().get_customer_id()).collect()
+}
+```
+
+The versions of `unique_ids` shown here implement the *concept* of a lens.
+
+Lenses are a way to allow accessing parts of a data type in an abstract,
+unified way.[^1]
+Lenses define a generic operation -- in our case, getting a customer ID from a
+record -- that can apply to multiple types.
+This is so even though each type has a different way to "observe" the desired
+data.
+
+At first glance, it may seem more like Rust is using "duck typing", the way
+that dynamic languages do.
+However, there are significant differences:
+
+1. Static type checking is maintained. Only types which "opt-in" to having a
+   customer ID (with an `impl CustomerId` block) can be placed into the
+   iterator or collection. Duck typed languages require this to be a runtime
+   check, which can then go wrong at runtime.
+1. The `impl CustomerId` block lives with the type. This means the type can
+   live in a different crate than definition of `CustomerId` itself,
+   maintaining extensibility.
+1. The definition of `CustomerId` is consistent throughout the program. Unlike
+   some approaches functional languages take (discussed below), it is
+   impossible to accidentally change the type in such a way the trait becomes
+   "invalid" at runtime.
+1. In many functional languages, data structures are immutable. While Rust's
+   traits allow mutation of types, the lens concept is designed to make deep
+   copying types with small changes much easier. (Deep copies which are often
+   optimized differently by the runtime. but that's another story.)
+
+Most functional languages take a different approach to this concept, however.
+They prefer instead to use other features Rust does not have -- a form of
+partial function construction known as "currying" -- to do it with functions
+instead.
+They "partially construct" the function, leaving the type of the object being
+operated on until the function is constructed, which Rust cannot do.
+
+This is the nearest thing to a Rust equivalent to the "pure functional way":
+
+```rust
+use std::any::Any;
+use std::collections::HashSet;
+
+pub trait CustomerId {
+    fn get_customer_id(&self) -> u64;
+}
+
+pub struct CreditRecord {
+    customer_id: u64,
+    name: String,
+    dob: String,
+    // other fields omitted
+}
+
+pub struct Account {
+    account_id: u32,
+    account_type: String,
+    // other fields omitted
+}
+
+pub struct AccountRecord {
+    primary_customer_id: u64,
+    accounts: Vec<Account>,
+}
+
+// the type of the lens itself, which is currently just a type alias unlike
+// functional languages
+/*
+type LenseFn<T> = dyn FnOnce(&dyn RecordType) -> Option<T>;
+*/
+// types that the lens can "view":
+/*
+trait RecordType {
+    fn view<T, L>(&self, lens: L) -> T
+        where L: FnOnce(&dyn RecordType) -> Option<T>
+    {
+        (lens)(self)
+    }
+}
+*/
+// all this is commented out because Rust does not support generics in traits
```

Comment:
Rust _does_ support generics in traits. You can declare generic functions in traits. And recently a subset of GATs made it into stable. You cannot use generics with _dynamic dispatch_. But you can even work around that one (ironically by using even more dynamic dispatching).